src/metabase/actions/models.clj

```diff
@@ -330,22 +332,122 @@
-      (m/assoc-some dashcard :action (get actions-by-id (:action_id dashcard))))))
+;; to avoid headaches with rewriting too much frontend code we will hack in a representation an operation and a single
+;; parameter as an integer. The frontend picker can when creating the dashcard pass the negative action_id.
+;; we will use an integer with an op (15 bits) and param (32 bits)
+;; to start with table actions, the param is the table-id.
+(def ^:private primitive-action-kind->int {:table.row/create 0 :table.row/update 1 :table.row/delete 2})
+(def ^:private int->primitive-action-kind (set/map-invert primitive-action-kind->int))
+
+(defn unpack-primitive-action-id
+  "Return an [op param] vector given the negative action id."
+  [^long encoded-id]
+  (let [pos-id     (bit-and (Math/abs encoded-id) 0xFFFFFFFFFFFF)
+        param-bits (bit-and pos-id 0xFFFFFFFF)
+        op-bits    (bit-and (bit-shift-right pos-id 32) 0xFFFF)]
+    [(int->primitive-action-kind op-bits) param-bits]))
+
+(defn- primitive-action-id ^long [kind ^long param]
+  (let [op-bits (primitive-action-kind->int kind)
+        packed  (bit-or (bit-shift-left op-bits 32) (bit-and param 0xFFFFFFFF))]
+    (- packed)))
+
+(defn table-primitive-action
+  "Return an action map for a table edit action for the primitive action kind.
+  Such an action is only valid if data editing is enabled for the database."
+  [table fields kind]
+  (let [field-param? (case kind
+                       :table.row/create
+                       #(not (:database_is_auto_increment %))
+                       :table.row/delete
+                       #(= :type/PK (:semantic_type %))
+                       (constantly true))
+        field-params (->> fields
+                          (filter field-param?)
+                          (sort-by :position))]
+    {:database_id (:database_id table)
+     :name        (name kind)
+     :kind        (str (namespace kind) "/" (name kind))
+     :table_id    (:id table)
+     :id          (primitive-action-id kind (:id table))
+     ;; true for all databases with data_editing_enabled
+     ;; it is expected this fn will not be called if this is not the case
+     :database_enabled_actions true
+     :visualization_settings
+     {:name ""
+      :type "button"
+      :description ""
+      :fields
+      (->> (for [field field-params
+                 :let [field-name (:name field)]]
+             [field-name
+              {:hidden      false
+               :id          field-name}])
+           (into {}))}
+     :parameters
+     (->> (for [field field-params
+                :let [field-name (:name field)]]
+            {:id    field-name
+             :display-name (:display_name field)
+             :type (:base_type field)
+             :target [:variable [:template-tag field-name]]
+             :slug  field-name
+             ;; todo
+             :required (or (= :type/PK (:semantic_type field))
+                           (:database_required field))
```

Comment:
We can just use ` database_is_nullable`, `database_is_generated`, and `database_default` right? If we don't already have a predicate for this we should probably extract one from wherever else we have this logic.